File: servers/deno/hub.ts
import type { RSession } from "./r_session.ts";

/** Placeholder for browser clients (implemented in be2.2). */
export interface BrowserClient {
  send(data: string): void;
  close(): void;
}

/**
 * Hub routes messages between R sessions and browser clients.
 * JS is single-threaded so no mutex is needed — Map/Set suffice.
 */
export class Hub {
  sessions = new Map<string, RSession>();
  clients = new Set<BrowserClient>();
  /** Maps metrics request ID → R session ID for routing responses. */
  metricsRouting = new Map<number, string>();
  verbose = false;

  registerSession(session: RSession): void {
    this.sessions.set(session.id, session);
    console.error(
      `R session registered: ${session.id} (total: ${this.sessions.size})`,
    );
  }

  unregisterSession(id: string): void {
    this.sessions.delete(id);
    console.error(
      `R session unregistered: ${id} (total: ${this.sessions.size})`,
    );
  }

  /**
   * Update a session's ID (when the real sessionId is extracted from the
   * first frame message).
   */
  updateSessionId(oldId: string, newId: string, session: RSession): void {
    this.sessions.delete(oldId);
    this.sessions.set(newId, session);
  }

  /** Broadcast a message string to all connected browser clients. */
  broadcastToClients(data: string): void {
    for (const client of this.clients) {
      try {
        client.send(data);
      } catch {
        // Slow/dead client — ignore
      }
    }
  }

  /** Broadcast a message string to all connected R sessions. */
  broadcastToR(data: string): void {
    for (const session of this.sessions.values()) {
      try {
        session.send(data);
      } catch (e) {
        console.error(
          `failed to send to R session ${session.id}: ${e}`,
        );
      }
    }
  }

  /**
   * Process a message from an R session.
   * Routes based on message type (frame, metrics_request, close, etc.).
   */
  handleRMessage(session: RSession, line: string): void {
    const type = extractType(line);

    switch (type) {
      case "frame": {
        // Inject sessionId into the plot object if not present
        let data = line;
        if (session.id && !line.includes('"sessionId"')) {
          data = injectSessionId(line, session.id);
        }
        this.broadcastToClients(data);
        if (this.verbose) {
          console.error(
            `frame from R session ${session.id} (${data.length} bytes)`,
          );
        }
        break;
      }

      case "metrics_request":
        this.handleMetricsRequest(session, line);
        break;

      case "close":
        if (this.verbose) {
          console.error(`device close from R session ${session.id}`);
        }
        this.broadcastToClients(line);
        break;

      default:
        // Unknown message type — forward to browsers
        this.broadcastToClients(line);
        break;
    }
  }

  /**
   * Route a metrics request from R to browsers, with timeout fallback.
   */
  private handleMetricsRequest(session: RSession, line: string): void {
    let id: number;
    try {
      const msg = JSON.parse(line);
      id = msg.id;
    } catch {
      console.error("failed to parse metrics request");
      return;
    }

    // No browsers connected → immediately send zero-value fallback
    if (this.clients.size === 0) {
      const fallback = JSON.stringify({
        type: "metrics_response",
        id,
        width: 0,
        ascent: 0,
        descent: 0,
      });
      try {
        session.send(fallback);
      } catch (e) {
        console.error(
          `failed to send metrics fallback to R session ${session.id}: ${e}`,
        );
      }
      return;
    }

    // Store routing: requestID → sessionID
    this.metricsRouting.set(id, session.id);

    // Forward to browsers
    this.broadcastToClients(line);

    // Timeout: if no response in 2s, send zero-value fallback
    setTimeout(() => {
      if (this.metricsRouting.has(id)) {
        this.metricsRouting.delete(id);
        const fallback = JSON.stringify({
          type: "metrics_response",
          id,
          width: 0,
          ascent: 0,
          descent: 0,
        });
        const target = this.sessions.get(session.id);
        if (target) {
          try {
            target.send(fallback);
          } catch (e) {
            console.error(
              `failed to send metrics fallback to R session ${session.id}: ${e}`,
            );
          }
        }
        if (this.verbose) {
          console.error(
            `metrics timeout for request ${id}, sent fallback to session ${session.id}`,
          );
        }
      }
    }, 2000);
  }

  /**
   * Route a metrics response from a browser to the originating R session.
   */
  handleMetricsResponse(line: string): void {
    let id: number;
    try {
      const msg = JSON.parse(line);
      id = msg.id;
    } catch {
      console.error("failed to parse metrics response");
      return;
    }

    const sessionId = this.metricsRouting.get(id);
    if (sessionId === undefined) {
      // Already timed out or duplicate
      return;
    }
    this.metricsRouting.delete(id);

    const session = this.sessions.get(sessionId);
    if (session) {
      try {
        session.send(line);
      } catch (e) {
        console.error(
          `failed to send metrics response to R session ${sessionId}: ${e}`,
        );
      }
    }
  }

  /** Register a browser client. */
  registerClient(client: BrowserClient): void {
    this.clients.add(client);
    console.error(
      `browser client connected (total: ${this.clients.size})`,
    );
  }

  /** Unregister a browser client. */
  unregisterClient(client: BrowserClient): void {
    this.clients.delete(client);
    console.error(
      `browser client disconnected (total: ${this.clients.size})`,
    );
  }

  /** Shut down the hub and close all connections. */
  close(): void {
    for (const client of this.clients) {
      try {
        client.close();
      } catch { /* ignore */ }
    }
    this.clients.clear();

    for (const session of this.sessions.values()) {
      try {
        session.close();
      } catch { /* ignore */ }
    }
    this.sessions.clear();
  }
}

/**
 * Extract the "type" field from a JSON line without full parse.
 * Falls back to empty string on malformed input.
 */
function extractType(line: string): string {
  const m = line.match(/"type"\s*:\s*"([^"]+)"/);
  return m ? m[1] : "";
}

/**
 * Inject sessionId into the plot object of a frame message.
 * Finds `"plot":{` or `"plot": {` and inserts `"sessionId":"<id>",` after
 * the opening brace — matching the Go server's injection logic.
 */
export function injectSessionId(line: string, sessionId: string): string {
  // Find "plot":{ or "plot": {
  const plotRe = /"plot"\s*:\s*\{/;
  const match = plotRe.exec(line);
  if (!match) return line;

  const insertPos = match.index + match[0].length;
  const escaped = JSON.stringify(sessionId);
  return (
    line.slice(0, insertPos) +
    `"sessionId":${escaped},` +
    line.slice(insertPos)
  );
}

File: servers/go/hub.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"sync"
	"time"
)

// Hub routes messages between R sessions and browser clients.
type Hub struct {
	mu               sync.RWMutex
	sessions         map[string]*RSession
	clients          map[*BrowserClient]bool
	metricsRouting   sync.Map // requestID(int) -> sessionID(string)
	registerClient   chan *BrowserClient
	unregisterClient chan *BrowserClient
	broadcastFrame   chan []byte
	done             chan struct{}
	runWg            sync.WaitGroup // tracks Run() goroutine
}

func NewHub() *Hub {
	h := &Hub{
		sessions:         make(map[string]*RSession),
		clients:          make(map[*BrowserClient]bool),
		registerClient:   make(chan *BrowserClient),
		unregisterClient: make(chan *BrowserClient),
		broadcastFrame:   make(chan []byte, 256),
		done:             make(chan struct{}),
	}
	h.runWg.Add(1) // Must be done before go hub.Run()
	return h
}

// Run is the main event loop. Must be launched in its own goroutine after NewHub().
func (h *Hub) Run() {
	defer h.runWg.Done()
	for {
		select {
		case client := <-h.registerClient:
			h.mu.Lock()
			h.clients[client] = true
			h.mu.Unlock()
			log.Printf("browser client connected (total: %d)", len(h.clients))

		case client := <-h.unregisterClient:
			h.mu.Lock()
			if _, ok := h.clients[client]; ok {
				delete(h.clients, client)
				close(client.send)
			}
			h.mu.Unlock()
			log.Printf("browser client disconnected (total: %d)", len(h.clients))

		case frame := <-h.broadcastFrame:
			h.mu.RLock()
			for client := range h.clients {
				select {
				case client.send <- frame:
				default:
					// Slow client, drop message
				}
			}
			h.mu.RUnlock()

		case <-h.done:
			return
		}
	}
}

// RegisterSession adds an R session to the hub.
func (h *Hub) RegisterSession(s *RSession) {
	h.mu.Lock()
	defer h.mu.Unlock()
	h.sessions[s.id] = s
	log.Printf("R session registered: %s (total: %d)", s.id, len(h.sessions))
}

// UnregisterSession removes an R session from the hub.
func (h *Hub) UnregisterSession(id string) {
	h.mu.Lock()
	defer h.mu.Unlock()
	delete(h.sessions, id)
	log.Printf("R session unregistered: %s (total: %d)", id, len(h.sessions))
}

// BroadcastToClients sends a message to all browser clients.
func (h *Hub) BroadcastToClients(data []byte) {
	h.mu.RLock()
	defer h.mu.RUnlock()
	for client := range h.clients {
		select {
		case client.send <- data:
		default:
		}
	}
}

// BroadcastToR sends a message to all R sessions.
func (h *Hub) BroadcastToR(data []byte) {
	h.mu.RLock()
	defer h.mu.RUnlock()
	for _, session := range h.sessions {
		if err := session.Send(data); err != nil {
			log.Printf("failed to send to R session %s: %v", session.id, err)
		}
	}
}

// BroadcastResizeToR sends a resize message to all R sessions,
// marking each session so the next frame can be tagged as a resize response.
// Duplicate resizes with identical dimensions are forwarded to R but do not
// re-arm the flag, preventing browser-side ws.onopen + ResizeObserver
// duplicates from tagging an extra frame.
func (h *Hub) BroadcastResizeToR(data []byte) {
	var dims struct {
		Width  int32 `json:"width"`
		Height int32 `json:"height"`
	}
	json.Unmarshal(data, &dims) // best-effort; zero on parse failure

	h.mu.RLock()
	defer h.mu.RUnlock()
	for _, session := range h.sessions {
		// Only increment when dimensions actually changed, so duplicate
		// resizes (ws.onopen + ResizeObserver with same dims) don't
		// cause extra frames to be tagged.
		oldW := session.lastResizeW.Load()
		oldH := session.lastResizeH.Load()
		if dims.Width != oldW || dims.Height != oldH {
			session.resizePending.Store(true)
			session.lastResizeW.Store(dims.Width)
			session.lastResizeH.Store(dims.Height)
		}
		if err := session.Send(data); err != nil {
			log.Printf("failed to send to R session %s: %v", session.id, err)
		}
	}
}

// msgType extracts the "type" field from a JSON message.
func msgType(data []byte) string {
	var msg struct {
		Type string `json:"type"`
	}
	if json.Unmarshal(data, &msg) != nil {
		return ""
	}
	return msg.Type
}

// HandleRMessage processes a message from an R session.
func (h *Hub) HandleRMessage(session *RSession, data []byte) {
	switch msgType(data) {
	case "frame":
		// Tag resize-triggered frames so the browser can update in place
		if session.resizePending.CompareAndSwap(true, false) {
			data = injectResizeFlag(data)
		}
		// Inject sessionId into the plot object if not present
		if session.id != "" && !bytes.Contains(data, []byte(`"sessionId"`)) {
			data = injectSessionID(data, session.id)
		}
		select {
		case h.broadcastFrame <- data:
		default:
		}
		if verbose {
			log.Printf("frame from R session %s (%d bytes)", session.id, len(data))
		}

	case "metrics_request":
		h.handleMetricsRequest(session, data)

	case "close":
		if verbose {
			log.Printf("device close from R session %s", session.id)
		}
		h.BroadcastToClients(data)

	default:
		// Unknown message type, forward to browsers
		h.BroadcastToClients(data)
	}
}

// handleMetricsRequest routes a metrics request from R to browsers,
// setting up a timeout for fallback.
func (h *Hub) handleMetricsRequest(session *RSession, data []byte) {
	// Extract request ID
	var msg struct {
		ID int `json:"id"`
	}
	if err := json.Unmarshal(data, &msg); err != nil {
		log.Printf("failed to parse metrics request: %v", err)
		return
	}

	// If no browsers are connected, immediately send zero-value fallback.
	// R treats zero-value responses the same as a timeout: it falls through
	// to local font metric approximations, but without the 500ms poll wait.
	//
	// Note: there is a benign TOCTOU race — a browser could connect between
	// the unlock and the send. The only consequence is one request using
	// local approximation instead of browser-measured metrics, which is
	// identical to the 2s timeout fallback behavior.
	h.mu.RLock()
	nClients := len(h.clients)
	h.mu.RUnlock()

	if nClients == 0 {
		fallback := fmt.Sprintf(`{"type":"metrics_response","id":%d,"width":0,"ascent":0,"descent":0}`, msg.ID)
		if err := session.Send([]byte(fallback)); err != nil {
			log.Printf("failed to send metrics fallback to R session %s: %v", session.id, err)
		}
		return
	}

	// Store routing: requestID -> sessionID
	h.metricsRouting.Store(msg.ID, session.id)

	// Forward to browsers
	h.BroadcastToClients(data)

	// Set timeout: if no response in 2s, send zero-value fallback
	time.AfterFunc(2*time.Second, func() {
		if _, loaded := h.metricsRouting.LoadAndDelete(msg.ID); loaded {
			// No response received, send fallback
			fallback := fmt.Sprintf(`{"type":"metrics_response","id":%d,"width":0,"ascent":0,"descent":0}`, msg.ID)
			if err := session.Send([]byte(fallback)); err != nil {
				log.Printf("failed to send metrics fallback to R session %s: %v", session.id, err)
			}
			if verbose {
				log.Printf("metrics timeout for request %d, sent fallback to session %s", msg.ID, session.id)
			}
		}
	})
}

// HandleMetricsResponse routes a metrics response from a browser to the originating R session.
func (h *Hub) HandleMetricsResponse(data []byte) {
	var msg struct {
		ID int `json:"id"`
	}
	if err := json.Unmarshal(data, &msg); err != nil {
		log.Printf("failed to parse metrics response: %v", err)
		return
	}

	val, loaded := h.metricsRouting.LoadAndDelete(msg.ID)
	if !loaded {
		// Already timed out or duplicate
		return
	}

	sessionID, ok := val.(string)
	if !ok {
		return
	}

	h.mu.RLock()
	session, exists := h.sessions[sessionID]
	h.mu.RUnlock()

	if exists {
		if err := session.Send(data); err != nil {
			log.Printf("failed to send metrics response to R session %s: %v", sessionID, err)
		}
	}
}

// Close shuts down the hub and all connections.
// It stops Run() first and waits for it to exit, then cleans up.
func (h *Hub) Close() {
	close(h.done)

	// Drain remaining channel messages so Run() can exit if it's
	// blocked trying to send on registerClient/unregisterClient.
	// Once done is closed, Run() returns on next select iteration.
	// Give it a moment to exit before we clean up.
	h.runWg.Wait()

	h.mu.Lock()
	defer h.mu.Unlock()

	for client := range h.clients {
		close(client.send)
		client.conn.Close()
		delete(h.clients, client)
	}

	for id, session := range h.sessions {
		session.conn.Close()
		delete(h.sessions, id)
	}
}

// injectResizeFlag adds "resize":true to a frame message so the browser
// knows this frame is a response to a resize event, not a new plot.
func injectResizeFlag(data []byte) []byte {
	idx := bytes.IndexByte(data, '{')
	if idx < 0 {
		return data
	}
	insert := []byte(`"resize":true,`)
	result := make([]byte, 0, len(data)+len(insert))
	result = append(result, data[:idx+1]...)
	result = append(result, insert...)
	result = append(result, data[idx+1:]...)
	return result
}

// injectSessionID adds sessionId to the plot object in a frame message.
func injectSessionID(data []byte, sessionID string) []byte {
	// Find "plot":{  and insert "sessionId":"..." after the opening brace
	plotIdx := bytes.Index(data, []byte(`"plot":{`))
	if plotIdx < 0 {
		plotIdx = bytes.Index(data, []byte(`"plot": {`))
		if plotIdx < 0 {
			return data
		}
	}
	braceIdx := bytes.IndexByte(data[plotIdx:], '{')
	if braceIdx < 0 {
		return data
	}
	insertPos := plotIdx + braceIdx + 1

	// JSON-encode the session ID to prevent injection
	escaped, err := json.Marshal(sessionID)
	if err != nil {
		return data
	}
	insert := append([]byte(`"sessionId":`), escaped...)
	insert = append(insert, ',')

	result := make([]byte, 0, len(data)+len(insert))
	result = append(result, data[:insertPos]...)
	result = append(result, insert...)
	result = append(result, data[insertPos:]...)
	return result
}
